### ISSUE TITLE
Image Preprocessing Difference Between Predictor and Validator

### ISSUE THREAD

**Issue by garricklw** [ultralytics/ultralytics#2531]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar bug report.

### YOLOv8 Component

Detection

### Bug

The issue I was running into was I was getting wildly different bounding boxes and confidences when calling model.val() and model.predict() with the exact same data files and the same hyperparameters.

When debugging and stepping through, I noticed that the image that was being passed to model() in validator.py:158 and predictor.py:230 were different sizes. This resulted in different predictions and thus different results.

My inputs are .jpg files, saved in RGB format, 595 x 128 x 3 (RGB channel dimension). 

When I run model.val() I pass the path to my dataset.yaml file, my conf=0.34, iou=0.7, and imgsz=(595, 128)
In profiling model.val I noticed that the image size before getting passed to self.preprocess(batch) had been changed to 3 x 160 x 640, the size difference being padded with grey pixels. 

When I ran model.predict by passing a path to my .jpg file, using the same hyperparameters, the results I got were completely different than model.val(). I noticed that the data loader was not padding the image, but was instead just calling cv2.imread() and passing that to the preprocess function.

When I manually loaded the .jpg file with cv2, padded the image up to 160 x 640 x 3 with grey pixels, and changed imgsz to (160, 640) I saw similar results to model.val(), still confirming if they're exactly the same results.

I think the fix is simple and logical: please expose the exact same data loading and preprocessing capabilities to the .predict method that are available to the .val method. Either allowing predict to accept a dataset.yaml or just making the .jpg file path reader work the same as the .yaml reader in .predict would make sense.

Everything else about the interface has been so easy to use, I was kind of shocked this issue was as insidious as it was.

### Environment

 - YOLO: Ultralytics YOLOv8.0.92 🚀 Python-3.10.10 torch-2.0.0 CPU
 - OS: Trained on Google Colab, Predicting on MacOS 13.2.1 (22D68)

### Minimal Reproducible Example

```
dataset_dir = local_data_dir / "nist_dataset_specgrams_debug" / "valid"
label_dir = dataset_dir / "labels"
experiment_dir = results_dir / "yolov8_nist" / "v8l_3000nist"
out_dir = experiment_dir / "valid"
os.makedirs(out_dir, exist_ok=True)

weights_file = experiment_dir / "train" / "weights" / "best.pt"
model = YOLO(weights_file)
# this gives expected results, images that get outputted look correct and bounding boxes are good
metrics = model.val(data=str(dataset_dir.parent / 'nist_dataset.yaml'), conf=0.34, iou=0.7, imgsz=(595, 128))

# this call gives garbage results
results: List[Results] = model.predict(source=str(dataset_dir / "images" / "example_2726.jpg"), conf=0.4, iou=0.7, imgsz=(595, 128), stream=False) 

# loading manually and padding gives much better results
source = cv2.imread(str(dataset_dir / "images" / "example_2726.jpg"))
x_target, y_target = 160, 640
x_diff, y_diff = 160 - 128, 640 - 595

x_pad = np.tile(114, reps=(x_diff, 595, 3))
source = np.concatenate([source, x_pad], axis=0)

y_pad = np.tile(114, reps=(160, y_diff, 3))
source = np.concatenate([source, y_pad], axis=1)

cv2.imwrite(str(dataset_dir / "images" / "example_2726_reshaped.jpg"), source)

results: List[Results] = model.predict(source=source, conf=0.34, iou=0.7, imgsz=(160, 640), stream=False) 
```

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by garricklw** [comment]
I updated to YOLOv8.0.98 and now I'm also seeing a discrepancy between the results output from the .train() call and the .val() and .predict(). That is, the images and plots look completely normal in .train(), but when I call .val() or .predict() I get garbage bounding boxes and confidences. 

I'm going to try padding all of my images up to 640 x 640 and see if it works, but it seems like there are some issues with arbitrary image size support.

**Comment by glenn-jocher** [comment]
@garricklw thank you for raising this issue with YOLOv8 and sorry for the inconvenience caused! 

It is possible that the issue may be due to changes in input image size or padding during data loading. It may be worth trying to resize and/or pad all of your input images to 640x640, which can be set as your `imgsz` argument when calling these methods. 

In any case, if the issue persists, don't hesitate to provide us with a minimum reproducible example and other relevant information such as your environment specifications for us to better assist you. We are dedicated to resolving issues promptly and consistently improving YOLOv8.

**Comment by garricklw** [comment]
Everything did work when training on 640 x 640, however this isn't an ideal solution, so I'll post a minimum reproducible example below with files attached to try to work towards the imgsz argument working as expected

**Comment by garricklw** [comment]
The model was trained with the following code: 

```
from ultralytics import YOLO

effort_name = "yolov8_nist"
experiment_name = "min_reproducible"

model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

model.train(data=str(dataset_dir / 'nist_dataset.yaml'), epochs=200, patience=10, optimizer="Adam", imgsz=(595, 128), exist_ok=True)

Example image and label from my dataset, image size (w, h): (595, 128)
[example_2841.txt](https://github.com/ultralytics/ultralytics/files/11490593/example_2841.txt)
![example_2841](https://github.com/ultralytics/ultralytics/assets/52170140/f80387e9-d21e-4776-b340-f220e2ce996c)

Training results from model: validation and training are working as expected
[train.zip](https://github.com/ultralytics/ultralytics/files/11490618/train.zip)

```

**Comment by garricklw** [comment]
Below is a cleaner example showing the issue: .val() and .predict() on the exact same image file with the exact same hyperparameters return completely different results.

```
from typing import List

from ultralytics import YOLO
from ultralytics.yolo.engine.results import Results

from perc_rf.config import results_dir, local_data_dir

if __name__ == '__main__':
    n_classes = 5

    # use the path to the attached dataset directory
    dataset_dir = local_data_dir / "nist_dataset_specgrams_debug" / "valid"
    label_dir = dataset_dir / "labels"

    experiment_dir = results_dir / "yolov8_nist" / "min_reproducible"
    # weight file is in train.zip in a previous comment
    weights_file = experiment_dir / "train" / "weights" / "best.pt"

    model = YOLO(weights_file)
    # Validation results are attached, everything works as expected
    metrics = model.val(data=str(dataset_dir.parent / 'nist_dataset.yaml'), conf=0.01, iou=0.7, imgsz=(595, 128))

    # using the same images as validation
    source = dataset_dir / "images"
    # this is the issue right now. Using the exact same hyperparameters, we get different results.
    # that is: 3 much less confident bounding boxes rather than one confident one.
    results: List[Results] = model.predict(source=source, conf=0.01, iou=0.7, imgsz=(595, 128), stream=False)  # list of Results objects
    
    # validation results are correct (100% precision and recall)
    print(metrics.results_dict)
    
    # .predict results are bogus, three boxes with low confidence
    print(len(results[0].boxes))
    print(results[0].boxes.conf)
    print(results[0].boxes.cls)
```

**Comment by garricklw** [comment]
Actually I think I may have just stumbled on the solution, and it should be a really easy fix:

When the code reaches Line 158 of validator.py, the img size passed to the model is 1 x 3 x 160 x 640. 

When the code reaches Line 240 of predictor.py, the img size passed to the model is 1 x 3 x 32 x 128.

HOWEVER, when I flipped the imgsz hyperparameter to (H, W), (128, 595) the img size being passed was now 1 x 3 x 128 x 608 and the results I got back were correct.

So, it looks like there is a bug in the predict function that flips the width and height of the imgsz argument. Note that in the docs: https://docs.ultralytics.com/modes/train/#arguments, it says the argument should be (w, h)

I suspect it's because Open-CV uses h, w, c ordering. I do believe that your interface (imgsz) should stay the same regardless of what the underlying image reading framework is, so I believe it is currently a bug that (h, w) needs to be passed for image reading approaches that use Open-CV.

**Comment by glenn-jocher** [comment]
@garricklw thank you for reporting this issue with YOLOv8. We appreciate your detailed analysis and debugging!

It appears that the issue is related to the order of the `imgsz` parameter when using the `predict()` function. The current behavior of the `predict()` function is to switch the order of the input dimensions from (h, w) to (w, h), which causes the incorrect image size to be passed to the model. This behavior is different from the `train()` and `val()` functions, which correctly pass the image size as (h, w).

We agree with your suggestion that the `imgsz` parameter should be consistent across all functions, regardless of the underlying image reading framework. We will investigate this issue further and make the necessary corrections in an upcoming release of YOLOv8.

In the meantime, as you have discovered, a possible workaround is to manually flip the order of the `imgsz` parameter in the `predict()` function from (h, w) to (w, h), or to use an image reading framework other than OpenCV that does not require this flipping.

Thank you for bringing this issue to our attention and please let us know if you have any further questions or concerns.
